go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2169,6 +2170,368 @@ func (s *VtctldServer) UpdateCellsAlias(ctx context.Context, req *vtctldatapb.Up
 	}, nil
 }
 
+// Validate is part of the vtctlservicepb.VtctldServer interface.
+func (s *VtctldServer) Validate(ctx context.Context, req *vtctldatapb.ValidateRequest) (*vtctldatapb.ValidateResponse, error) {
+	span, ctx := trace.NewSpan(ctx, "VtctldServer.Validate")
+	defer span.Finish()
+
+	span.Annotate("ping_tablets", req.PingTablets)
+
+	resp := vtctldatapb.ValidateResponse{}
+	getKeyspacesCtx, getKeyspacesCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+	defer getKeyspacesCancel()
+
+	keyspaces, err := s.ts.GetKeyspaces(getKeyspacesCtx)
+	if err != nil {
+		resp.Results = append(resp.Results, fmt.Sprintf("GetKeyspaces failed: %v", err))
+		return &resp, nil
+	}
+
+	var (
+		m  sync.Mutex
+		wg sync.WaitGroup
+	)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		validateAllTablets := func(ctx context.Context, keyspaces []string) {
```

Comment:
Another deviation from the original implementation: we used to make a call to `GetKeyspaces` in the outer function, and then the first thing this goroutine would do is ....... also make a call to `GetKeyspaces`, so, I just removed it and used the earlier result.